### DISCUSSION TITLE
Action of residuals in chemical potential kernels for KKS

### DISCUSSION THREAD

**Discussion by aclin4** [idaholab/moose#21666] [Q&A Modules: Phase field] [answered]
I'm trying to understand how exactly the non-linear variables are updated in kernels such as `KKSPhaseChemicalPotential` or `SLKKSChemicalPotential`. From my basic understanding, non-linear variables are updated using the residual _R_ calculated in each kernel following:

c(new) = c(old) + R

However in the case of the chemical potential equality kernels, the non-linear variable only appears in the residual in the form of a partial derivative:

R = df/dc - df'/dc'

where c is the non-linear variable. In the kernels where the non-linear variable does not directly appear in the residual, how exactly is the calculated value of the residual used to update the non-linear variable?

Thanks!
Albert

**Comment by GiudGiud** [comment, marked as answer]
hello

```
c(new) = c(old) + R dt
```
this is only **very roughly (and actually not really)** true for first order Euler time integrating.

The variables are updated as the result of the nonlinear solve. We can handle many time integration schemes, we can handle nonlinear variables that dont appear in the equations directly (like hidden in material properties for example), etc. 
In the end, the solution returned when convergence is reached is the solution vector such that the residual is minimized.

Guillaume

**Reply by aclin4** [reply]
I see. That makes more sense now, thanks!
